### PR TITLE
Improve query plan when Hive tables has compatible bucket number

### PR DESCRIPTION
**Pull request by haozhun**
When two Hive tables have same bucketing key, but different bucket number,
remote exchange is not necessary to join them when the bucket numbers are
compatible.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 1]
```diff
@@ -127,6 +130,52 @@ private static SubPlan analyzeGroupedExecution(Session session, Metadata metadat
...
+        return reassignPartitioningHandleIfNecessaryHelper(session, metadata, subPlan, subPlan.getFragment().getPartitioning());
+    }
+
+    private SubPlan reassignPartitioningHandleIfNecessaryHelper(Session session, Metadata metadata, SubPlan subPlan, PartitioningHandle newOutputPartitioningHandle)
+    {
+        PlanFragment fragment = subPlan.getFragment();
+
+        PartitioningHandleReassigner partitioningHandleReassigner = new PartitioningHandleReassigner(fragment.getPartitioning(), metadata, session);
+        PlanNode newRoot;
+        if (fragment.getPartitioning().isSingleNode()) {
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: Should the condition here and the condition on 153 the same?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 2]
```diff
@@ -217,7 +266,16 @@ public PlanNode visitTableScan(TableScanNode node, RewriteContext<FragmentProper
                     .map(TablePartitioning::getPartitioningHandle)
                     .orElse(SOURCE_DISTRIBUTION);
 
-            context.get().addSourceDistribution(node.getId(), partitioning);
+            context.get().addSourceDistribution(node.getId(), partitioning, metadata, session);
+            return context.defaultRewrite(node, context.get());
+        }
+
+        @Override
+        public PlanNode visitTableWriter(TableWriterNode node, RewriteContext<FragmentProperties> context)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: Shall this go to a separate commit ? :)

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 3]
```diff
@@ -314,13 +314,13 @@ private void invokeNoMoreSplitsIfNecessary()
         if (partition.getPartition().isPresent()) {
```
Comment (kokosing) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: What does it mean `bucket numbers are compatible`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 4]
```diff
@@ -127,6 +130,52 @@ private static SubPlan analyzeGroupedExecution(Session session, Metadata metadat
...
+    {
+        return reassignPartitioningHandleIfNecessaryHelper(session, metadata, subPlan, subPlan.getFragment().getPartitioning());
+    }
+
+    private SubPlan reassignPartitioningHandleIfNecessaryHelper(Session session, Metadata metadata, SubPlan subPlan, PartitioningHandle newOutputPartitioningHandle)
+    {
+        PlanFragment fragment = subPlan.getFragment();
+
+        PartitioningHandleReassigner partitioningHandleReassigner = new PartitioningHandleReassigner(fragment.getPartitioning(), metadata, session);
+        PlanNode newRoot;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: initialize it to `fragment.getRoot()`, invert the condition in below if and remove else. Then it will follow the pattern on 152.

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java [hunk 5]
```diff
@@ -98,6 +98,20 @@ default boolean schemaExists(ConnectorSession session, String schemaName)
 
     ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTableLayoutHandle handle);
 
+    default ConnectorTableLayoutHandle getAlternativeLayoutHandle(ConnectorSession session, ConnectorTableLayoutHandle tableLayoutHandle, ConnectorPartitioningHandle partitioningHandle)
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java]: This surely missing a comment.

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java [hunk 6]
```diff
@@ -98,6 +98,20 @@ default boolean schemaExists(ConnectorSession session, String schemaName)
...
+    {
+        throw new PrestoException(GENERIC_INTERNAL_ERROR, "ConnectorMetadata getCommonPartitioningHandle() is implemented without getAlternativeLayout()");
+    }
+
+    default Optional<ConnectorPartitioningHandle> getCommonPartitioningHandle(ConnectorSession session, ConnectorPartitioningHandle left, ConnectorPartitioningHandle right)
+    {
+        // By default, two ConnectorPartitioningHandle are compatible if and only if they are equal
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java]: this is already written below in more or less the same amount of characters

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 7]
```diff
@@ -127,6 +130,52 @@ private static SubPlan analyzeGroupedExecution(Session session, Metadata metadat
         return new SubPlan(fragment, result.build());
     }
 
+    private SubPlan reassignPartitioningHandleIfNecessary(Session session, Metadata metadata, SubPlan subPlan)
+    {
+        return reassignPartitioningHandleIfNecessaryHelper(session, metadata, subPlan, subPlan.getFragment().getPartitioning());
+    }
+
+    private SubPlan reassignPartitioningHandleIfNecessaryHelper(Session session, Metadata metadata, SubPlan subPlan, PartitioningHandle newOutputPartitioningHandle)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: remove `Helper` part from the method name

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 8]
```diff
@@ -127,6 +130,52 @@ private static SubPlan analyzeGroupedExecution(Session session, Metadata metadat
...
+    {
+        PlanFragment fragment = subPlan.getFragment();
+
+        PartitioningHandleReassigner partitioningHandleReassigner = new PartitioningHandleReassigner(fragment.getPartitioning(), metadata, session);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: move it to the else block below

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 9]
```diff
@@ -586,4 +671,43 @@ public boolean isSubTreeUseful()
             return capableTableScanNodes;
         }
     }
+
+    private class PartitioningHandleReassigner
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: static final

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java (hunk 5 above)
Comment (kokosing) [reply presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java]: Maybe getPartitionedTableLayoutHandle? Maybe it could return Optional, then I suspect that below method would not need to be necessary.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 10]
```diff
@@ -1529,6 +1533,78 @@ public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTa
...
+            //TODO! commented out to run travis. uncomment before merge
+            //return Optional.empty();
+        }
+        int largerBucketCount = Math.max(leftHandle.getBucketCount(), rightHandle.getBucketCount());
+        int smallerBucketCount = Math.min(leftHandle.getBucketCount(), rightHandle.getBucketCount());
+        if (largerBucketCount % smallerBucketCount != 0) {
+            // must be evenly divisible
+            return Optional.empty();
+        }
+        if (Integer.bitCount(largerBucketCount / smallerBucketCount) != 1) {
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: I agree that this isn't necessary. But this is a reasonable restriction, and the existing bucket # evolution also has restriction.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 11]
```diff
@@ -1529,6 +1533,78 @@ public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTa
...
+            // ratio must be power of two
+            return Optional.empty();
+        }
+
+        OptionalInt maxCompatibleBucketCount = min(leftHandle.getMaxCompatibleBucketCount(), rightHandle.getMaxCompatibleBucketCount());
+        if (maxCompatibleBucketCount.isPresent() && maxCompatibleBucketCount.getAsInt() < smallerBucketCount) {
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: This is true. I suppose you are concerned about this.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 12]
```diff
@@ -1529,6 +1533,78 @@ public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTa
...
+        HivePartitioningHandle leftHandle = (HivePartitioningHandle) left;
+        HivePartitioningHandle rightHandle = (HivePartitioningHandle) right;
+        if (!leftHandle.getHiveTypes().equals(rightHandle.getHiveTypes())) {
+            return Optional.empty();
+        }
+        if (!isOptimizedMismatchedBucketCount(session) && leftHandle.getBucketCount() != rightHandle.getBucketCount()) {
+            //TODO! commented out to run travis. uncomment before merge
+            //return Optional.empty();
+        }
+        int largerBucketCount = Math.max(leftHandle.getBucketCount(), rightHandle.getBucketCount());
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: From the abstraction perspective: Even though both functions take two integers, the semantics is completely different. One takes table bucket count and partition bucket, the other takes two table bucket count (and order doesn't matter).

From the practical perspective: On the other hand, while the code looks similar, the logic is sufficiently different. While code reuse is still possible, it won't be particularly straightforward.

Overall, I think the two functions should be kept separate.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveTableLayoutHandle.java [hunk 13]
```diff
@@ -125,6 +125,11 @@ public SchemaTableName getSchemaTableName()
     @Override
     public String toString()
     {
-        return schemaTableName.toString();
+        StringBuilder result = new StringBuilder();
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveTableLayoutHandle.java]: `toString` of `ConnectorTableLayoutHandle` is used for EXPLAIN. As a result, using `toStringHelper` would lead to too much boilerplate.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ActualProperties.java [hunk 14]
```diff
@@ -393,24 +397,28 @@ private boolean isNodePartitionedOn(Collection<Symbol> columns, Set<Symbol> cons
             return nodePartitioning.isPresent() && nodePartitioning.get().isPartitionedOn(columns, constants) && this.nullsAndAnyReplicated == nullsAndAnyReplicated;
         }
 
-        private boolean isNodePartitionedOn(Partitioning partitioning, boolean nullsAndAnyReplicated)
+        private boolean isCompatibleTablePartitioningWith(Partitioning partitioning, boolean nullsAndAnyReplicated, Metadata metadata, Session session)
         {
-            return nodePartitioning.isPresent() && nodePartitioning.get().equals(partitioning) && this.nullsAndAnyReplicated == nullsAndAnyReplicated;
+            return nodePartitioning.isPresent() && nodePartitioning.get().isCompatibleWith(partitioning, metadata, session) && this.nullsAndAnyReplicated == nullsAndAnyReplicated;
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ActualProperties.java]: I believe this would be a reasonable rename to make. I don't think it fits in this PR though.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 15]
```diff
@@ -428,12 +429,35 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
 
         // convert files internal splits
         List<InternalHiveSplit> splitList = new ArrayList<>();
-        for (int bucketNumber = 0; bucketNumber < Math.max(tableBucketCount, partitionBucketCount); bucketNumber++) {
-            int partitionBucketNumber = bucketNumber % partitionBucketCount; // physical
-            int tableBucketNumber = bucketNumber % tableBucketCount; // logical
-            if (bucketSplitInfo.isBucketEnabled(tableBucketNumber)) {
+        for (int bucketNumber = 0; bucketNumber < Math.max(readBucketCount, partitionBucketCount); bucketNumber++) {
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: There is a desire here to make sure that the splits are produced in an order such that it round robins (1 split at a time for each "read bucket id").

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java (hunk 1 above)
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: I don't think so

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java (hunk 5 above)
Comment (haozhun) [reply presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java]: I added comment.

I don't think returning an Optional would be helpful in removing the other method.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java (hunk 7 above)
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: I prefer keeping the suffix as It is a helper method which is necessary in order to do recursion

File: presto-tests/src/main/java/com/facebook/presto/tests/QueryAssertions.java [hunk 16]
```diff
@@ -51,12 +56,26 @@ private QueryAssertions()
...
+            queryPlan = resultWithPlan.getQueryPlan();
+            results = resultWithPlan.getMaterializedResult().toTestTypes();
+        }
+        else {
+            queryPlan = null;
+            results = queryRunner.execute(session, sql);
+        }
         log.info("FINISHED in presto: %s", nanosSince(start));
 
+        if (planAssertion.isPresent()) {
```
Comment (wenleix) [inline presto-tests/src/main/java/com/facebook/presto/tests/QueryAssertions.java]: We can also write as
```java
        planAssertion.ifPresent(assertion -> assertion.accept(queryPlan));
```

ditto for line 150.

I don't have strong option here, though. Given we anyway have to have the if-statement from line 64-72

Comment (haozhun) [reply presto-tests/src/main/java/com/facebook/presto/tests/QueryAssertions.java]: I'll leave this as is.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 17]
```diff
@@ -539,9 +539,18 @@ public int getBucketCount()
             return bucketCount;
         }
 
-        public boolean isBucketEnabled(int value)
+        /**
+         * Evaluates whether the provided table bucket number passes the bucket predicate.
+         * A bucket predicate can be present in two cases:
+         * <ul>
+         * <li>Filter on "$bucket" column. e.g. {@code "$bucket" between 0 and 100}
+         * <li>Single-value equality filter on all bucket columns. e.g. for a table with two bucketing columns,
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: See `HiveBucketing.getHiveBucket`

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 18]
```diff
@@ -428,12 +429,37 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
...
+                }
+            }
+
+            if (containsEligibleTableBucket && containsIneligibleTableBucket) {
+                throw new PrestoException(
+                        NOT_SUPPORTED,
+                        "The bucket filter cannot be satisfied. There are restrictions on the bucket filter when all the following is true: " +
+                                "1. a table has a different buckets count as at least one of its partitions that is read in this query; " +
+                                "2. the table has a different but compatible bucket number with another table in the query; " +
+                                "3. some buckets of the table is filtered out from the query, most likely using a filter on \"$bucket\". " +
```
[the remark this answers is not in the document]
Comment (haozhun) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: Yes, that is correct.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 19]
```diff
@@ -2244,6 +2244,78 @@ public void testPredicatePushDownToTableScan()
...
+                    "FROM\n" +
+                    "  test_mismatch_bucketing16\n" +
+                    "JOIN\n" +
+                    "  test_mismatch_bucketing32\n" +
+                    "ON key16=key32\n" +
+                    "JOIN\n" +
+                    "  test_mismatch_bucketingN\n" +
+                    "ON key16=keyN";
+
+            assertUpdate(withoutMismatchOptimization, writeToTableWithMoreBuckets, 15000, assertRemoteExchangesCount(4));
```
Comment (wenleix) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: What about adding a test case about `writeToTableWithNoBuckets` ? :)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 20]
```diff
@@ -315,34 +361,44 @@ public FragmentProperties setSingleNodeDistribution()
             return this;
         }
 
-        public FragmentProperties setDistribution(PartitioningHandle distribution)
+        public FragmentProperties setDistribution(PartitioningHandle distribution, Metadata metadata, Session session)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: What about call it `coalesceDistribution`?  Or in the comment mention it now will coalesce fragment distribution to a common one

**Comment by haozhun** [comment]
Merged #11749

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 21]
```diff
@@ -375,28 +431,39 @@ public FragmentProperties setCoordinatorOnlyDistribution()
             return this;
         }
 
-        public FragmentProperties addSourceDistribution(PlanNodeId source, PartitioningHandle distribution)
+        public FragmentProperties addSourceDistribution(PlanNodeId source, PartitioningHandle distribution, Metadata metadata, Session session)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: Now two methods will `coalesce` fragment's distribution:
- `setDistribution`
- `addSourceDistribution`

Maybe consider to have some common method to do `coalesce`. This might help reason the code ? (e.g. when is fragment coalesce happen)
